### PR TITLE
use column to order network modification

### PR DESCRIPTION
**Pull request by JacquesBorsenberger**
Signed-off-by: jacques borsenberger <jacques.borsenberger@rte-france.com>

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/entities/ModificationGroupEntity.java [hunk 1]
```diff
@@ -28,6 +36,12 @@
     @Column(name = "id")
     private UUID id;
 
+    @OneToMany(fetch = FetchType.LAZY)
+    @JoinColumn(name = "groupId", foreignKey = @ForeignKey(name = "group_id_fk_constraint"), nullable = false)
+    @Setter
+    @OrderColumn
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/entities/ModificationGroupEntity.java:42]: With this approach we have as many update as elements in the list when we create a list of modifications and when we insert a list of modifications
Is this really what we want ?

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 2]
```diff
@@ -93,10 +98,30 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
...
     }
 
+    @Transactional // To have all move in the same transaction (atomic)
+    public void moveModifications(UUID groupUuid, List<UUID> modifications, UUID before) {
+        var modificationGroupEntity = getModificationGroup(groupUuid);
+
+        Map<UUID, ModificationEntity> originalModifications = modificationRepository.findAllBaseByGroupId(groupUuid).stream()
+            .collect(Collectors.toMap(ModificationEntity::getId, Function.identity(), (x, y) -> y, LinkedHashMap::new));
+
+        if (!originalModifications.keySet().containsAll(modifications) || (before != null && !originalModifications.containsKey(before))) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:112]: `before` can be null ?
Perhaps `@NonNull`

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 3]
```diff
@@ -93,10 +98,30 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
...
+        }
+
+        List<ModificationEntity> modificationsToMove = modifications.stream().map(originalModifications::remove).collect(Collectors.toList());
+
+        List<ModificationEntity> newModificationList = new ArrayList<>(originalModifications.values());
+        int index =  before == null ? newModificationList.size() : newModificationList.indexOf(originalModifications.get(before));
+        newModificationList.addAll(index, modificationsToMove);
+
+        modificationGroupEntity.setModifications(newModificationList);
+    }
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:123]: Let's have all modifications after `before` updated in database

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 4]
```diff
@@ -109,94 +134,84 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
...
+    public ModificationEntity getModificationEntity(UUID groupUuid, UUID modificationUuid, ModificationType type) {
+        if (modificationRepository.existsByIdAndGroupId(modificationUuid, groupUuid)) {
+            ModificationEntity modification = modificationRepository.getOne(modificationUuid);
+            if (modification.getType().equals(type.name())) {
+                return modification;
+            }
+        }
+        throw new NetworkModificationException(MODIFICATION_NOT_FOUND, modificationUuid.toString());
+    }
+
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:157]: Perhaps only 
```
    @Transactional(readOnly = true)
    public ModificationInfos getModificationInfo(UUID groupUuid, UUID modificationUuid, ModificationType type) {
        return modificationRepository
            .findById(modificationUuid)
            .orElseThrow(() -> new NetworkModificationException(MODIFICATION_NOT_FOUND, modificationUuid.toString()))
            .toModificationInfos();
    }
```

and remove `getModificationInfo`

File: src/main/java/org/gridsuite/modification/server/entities/ModificationEntity.java [hunk 5]
```diff
@@ -28,19 +28,14 @@
 @Setter
 @Entity
 @Inheritance(strategy = InheritanceType.JOINED)
-@Table(name = "modification", indexes = {@Index(name = "modificationEntity_group_id_index", columnList = "group_id")})
+@Table(name = "modification")
```
[the remark this answers is not in the document]
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/modification/server/entities/ModificationEntity.java:31]: group id is not there for jpa

File: src/main/resources/db/changelog/changesets/changelog_2022-02-15T14:32:25Z.xml [hunk 6]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.10.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.10.xsd">
+    <changeSet author="borsenbergerjac (generated)" id="1644935557396-8">
+        <addColumn tableName="modification">
+            <column name="modifications_order" type="int4"/>
+        </addColumn>
+    </changeSet>
+    <changeSet author="borsenbergerjac (generated)" id="1644935557396-1">
```
[the remark this answers is not in the document]
Comment (JacquesBorsenberger) [reply src/main/resources/db/changelog/changesets/changelog_2022-02-15T14:32:25Z.xml]: removed unwanted changesets

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java (hunk 2 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:112]: before can be null (will add comment)

File: src/test/java/org/gridsuite/modification/server/ModificationRepositoryTest.java [hunk 7]
```diff
@@ -20,8 +20,10 @@
 import org.gridsuite.modification.server.entities.equipment.modification.BranchStatusModificationEntity;
 import org.gridsuite.modification.server.entities.equipment.modification.attribute.EquipmentAttributeModificationEntity;
 import org.gridsuite.modification.server.repositories.ModificationGroupRepository;
+import org.gridsuite.modification.server.repositories.ModificationRepository;
 import org.gridsuite.modification.server.repositories.NetworkModificationRepository;
 import org.gridsuite.modification.server.utils.*;
+import org.jetbrains.annotations.NotNull;
```
[the remark this answers is not in the document]
Comment (JacquesBorsenberger) [reply src/test/java/org/gridsuite/modification/server/ModificationRepositoryTest.java]: done

File: src/main/java/org/gridsuite/modification/server/repositories/ModificationRepository.java [hunk 8]
```diff
@@ -19,10 +19,10 @@
  */
 @Repository
 public interface ModificationRepository extends JpaRepository<ModificationEntity, UUID> {
-    @Query(value = "SELECT *, 0 AS clazz_ FROM modification WHERE group_id = ?1", nativeQuery = true)
+    @Query(value = "SELECT *, 0 AS clazz_ FROM modification WHERE group_id = ?1 order by modifications_order", nativeQuery = true)
     List<ModificationEntity> findAllBaseByGroupId(UUID uuid);
 
-    List<ModificationEntity> findAllByGroupId(UUID uuid);
+    @Query(value = "SELECT count(*)>0 FROM modification WHERE id = ?1 and group_id = ?2", nativeQuery = true)
```
[the remark this answers is not in the document]
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/modification/server/repositories/ModificationRepository.java]: done

File: src/main/java/org/gridsuite/modification/server/entities/ModificationGroupEntity.java (hunk 1 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/modification/server/entities/ModificationGroupEntity.java:42]: yes the other option is to use native sql, and only work for inserting

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**